Load remote notes. Display & filter by labels.

tomusdrw opened the pull request:
This PR reworks a bit the internal model for storing notes. It's required to support features outlined in #83 

1. On top of that I've implemented displaying and filtering by labels (no user-defined labels yet).
2. Importing notes is not overwritting the existing ones (Closes #76)
3. Editing the node is re-worked (no more `onBlur`)
4. We have the ability to load remote notes from URL (see `remoteNotesSources`; no UI to edit it yet)

I don't want to add much more to this PR to make it easier to review. So far the changes should mostly only affect how notes are displayed. The only new feature visible at once is filtering by labels (but the user will only see one label: "local").

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -95,122 +160,268 @@ function loadLegacyFromLocalStorage(): string | null {
+  const updateLocalNotes = useCallback((localNotes: TAnyNote[]) => {
+    setHistory((history) => [...history, localNotes].slice(-1 * HISTORY_STEPS_LIMIT));
+    setLocalNotes(localNotes);
+    saveToLocalStorage(localNotes);
```
please remove unnecessary data, such as `canMigrateTo` or `local/remote/imported` labels, before saving to local storage.

skoszuta reviewed (changes requested):
Generally looks good. There are couple of things I'd like to point out though:
- It feels messy to save notes along with data that can be inferred in runtime, such as:
  - local/remote/imported labels (unless you'd drop the source property)
  - canMigrateTo: this one especially since it serves no purpose at all in local storage
- I feel like we should introduce note spec versioning. its a reasonable moment to do so because we can treat all notes without a spec property as v2. if we dont do this now were in for quite a mess on next iterations.

The rest is in the comments. Cheers

tomusdrw reviewed (commented):
All points are pretty valid and should now be addressed.

Decided to change the note data model a bit. Cleaned it up from unnecessary stuff:
1. `pageNumber` was duplicated in `selectionStart`
2. `selectionString` was not really used and can be better deducted from the blocks (we don't need it as a safety measure any more imho).
3. Introduced an explicit container for notes (instead of storing just an array), since it avoid possible inconsistency in notes inside the array (what if I have a bunch of V2 and V3 notes mixed together in one array?) and I'm pretty sure it might come handy in the future when we need to attach more data there.

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -57,160 +91,352 @@ function isINote(arg: unknown): arg is INote {
+      (jsonStr: string, label: string) => {
+        let newNotes = [];
+        try {
+          newNotes = parseNotesFromJson(jsonStr, `${LABEL_IMPORTED}${label}`).notes;
```
If i understand correctly, imported v3 notes will never receive the "Imported" label

tomusdrw replied:
Yes.

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -1,52 +1,86 @@
+
+export enum NoteSource {
+  Local = 0,
+  Imported = 1,
```
`NoteSource.Imported` is never used anywhere

tomusdrw replied:
Removed

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -57,160 +91,352 @@ function isINote(arg: unknown): arg is INote {
+          notes: parsed.map((note) => convertNoteV2toV3(note, defaultLabel)),
+        };
+      }
+      throw new Error("Notes JSON should be an array.");
```
This `throw` breaks v2 imports

tomusdrw replied:
Fixed.

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -57,160 +91,352 @@ function isINote(arg: unknown): arg is INote {
-      link.setAttribute("href", `data:application/json;charset=utf-8,${encodeURIComponent(strNotes)}`);
-      link.setAttribute("download", `graypaper-notes-${new Date().toISOString()}.json`);
-      link.click();
+      const strNotes = JSON.stringify(localNotes);
```
You export the `localNotes` array without the wrapper therefore they're always treated as V2 when importing.

tomusdrw replied:
Fixed.

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -57,160 +91,352 @@ function isINote(arg: unknown): arg is INote {
+          const newSelection = await migrateSelection({ selectionStart, selectionEnd }, version, currentVersion);
+
+          if (!newSelection) {
+            return {
```
this is the exact same return as in line 394. Please merge them

tomusdrw replied:
Fixed.

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -57,160 +91,352 @@ function isINote(arg: unknown): arg is INote {
+          // TODO [ToDr] We can potentially cache that by re-using the migrations
+          // that were already done.
+          const { version, selectionStart, selectionEnd } = note;
+          const hash = await xxhash32(`${source}-${note.version}-${note.date}-${note.content}`);
```
i do believe the hash lib is an overkill since you can simply use that string (or any simple hash doable in js) as key for react components. up to you though, its fun to have nice hashes

tomusdrw replied:
Removed.

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -1,52 +1,86 @@
 }
 
-export type TAnyNote = IPointNote | IHighlightNote;
+export type INote = ISelectionParams & {
```
The naming here is a bit confusing. There's `INoteV2`, `INoteV3` and then `INote` which has an entirely different hierarchy of properties and serves a different purpose. It would be nice to indicate that:
- it's a decorated version of a raw note (you change naming the other way around, by changing INoteV2 or INoteV3 to something else)
- the selection params on this one are migrated/temporary.

tomusdrw replied:
Reworked.

skoszuta reviewed (changes requested):
I agree with the goals you're going after but:
- import/export is broken
- some parts seem unfinished/untested/unclear
Details in the comments

As a general side note the code seems a little convoluted for what it does and NotesProvider grew kinda large. That's not a blocker but just my impression.

Cheers

tomusdrw commented:
Reworked & refactored the code further.

- [ ] Extracted hooks from `NotesProvider`
- [ ] Moved local storage management and import/export functions to utils and added tests.

Let me know wdyt.

skoszuta commented on src/components/NotesProvider/NotesProvider.tsx:
```diff
@@ -95,122 +160,268 @@ function loadLegacyFromLocalStorage(): string | null {
+            n.source = NoteSource.Remote;
+            return n;
+          });
+          remoteNotes.push(...notes);
```
indeed there isn't a biome rule for it. it looks like the lead maintainer has an opinion on this topic: https://github.com/biomejs/biome/discussions/3#discussioncomment-7379356

skoszuta reviewed (approved):
Looking good! Nice call with adding the tests.

One thing i noticed is that when a note fails automatic migration we show it to the user without an ability to migrate it manually. If that wasn't intentional that's the only thing I'd consider fixing.

Other than that lgtm 🚀

tomusdrw commented:
> Looking good! Nice call with adding the tests.
> 
> One thing i noticed is that when a note fails automatic migration we show it to the user without an ability to migrate it manually. If that wasn't intentional that's the only thing I'd consider fixing.
> 
> Other than that lgtm 🚀

Good catch, fixed, thanks!
